Slice close-block (#3125 guard) fires on a consensus-reached slice after restart_phase → cascades to fail the whole phase (regression / role=unassigned orphan)

## Summary

A slice-DAG implement phase (`issue-3312`, build `42eb9511f`) **failed the entire phase** because a slice that had **already reached full BRC consensus** was close-blocked by the **#3125 guard** ("task records cite commits unreachable from the integration branch") — even though **#3125 is CLOSED/fixed**. This is a regression or an uncovered case. The trigger is a task record citing an **orphaned commit attributed to `role=unassigned`**, which appeared after a `restart_phase`.

## What happened

`issue-3312` runs the 19-file decomposition as a **strictly sequential 19-slice chain**. The implement phase was restarted once via `restart_phase` (the prior run parked on a quota streak-10). On the restarted run:

- **slice-1** (`contract_cli.py`) landed cleanly — `Consensus reached; phase complete`.
- **slice-2** (`routes/decisions.py`) reached full BRC consensus — all 8 agents confirmed; `04:11:24 event_loop: consensus complete, stopping slice_id=slice-2`.
- **Immediately after**, slice close was blocked:

```
04:11:27 [ERROR] orchestrator.pipelines: Slice close blocked: task records cite commits
  unreachable from the integration branch (#3125)
  pipeline_id=issue-3312 slice_id=slice-2 integration_branch=egg/issue-3312/slice-2
  unreachable="task-2-4 (role=unassigned, commit=e1cf4a48208f89d54a404...)"
```

- That marked slice-2 **failed**, which cascaded (correctly, for a sequential chain) to block all downstream slices and fail the phase:

```
04:12:27 [WARNING] Slice cascade fired failed_slice=slice-2 blocked=['slice-3',...,'slice-19']
04:12:27 [ERROR]   Phase failed pipeline_id=issue-3312 phase=implement exit_code=1
                   container_logs="--- slice slice-1 --- Consensus reached; phase complete. --- slice slice-2 --..."
```

## Why this matters

- **A recoverable bookkeeping mismatch nukes a multi-hour, multi-slice run.** slice-2's actual work reached consensus; the only problem is a task record (`task-2-4`) pointing at a commit not on the integration branch. Yet the whole phase fails.
- **#3125 is closed as fixed** — so either the fix regressed, or it doesn't cover this path. The distinguishing feature here is `role=unassigned` on the orphan commit and that it surfaced **after `restart_phase`**.

## Root-cause hypothesis

`restart_phase` deletes per-agent worktrees, best-effort salvages unpushed commits to `egg/recovered/*`, and re-forks fresh worktrees from the shared work-branch tip. A commit recorded in `task-2-4`'s record **before** the restart (or made by a since-reaped concurrent sibling producer — this pipeline also hit a concurrent-producer race, see #3337) is **not reachable** from the freshly re-forked `egg/issue-3312/slice-2` integration branch. The `role=unassigned` attribution is consistent with a salvaged/orphaned commit whose author-role mapping was lost. The #3125 close guard then fires on it.

So the gap is likely: **`restart_phase` (and/or concurrent-producer reaping) can leave task records citing commits that the post-restart integration branch can't reach**, and the #3125 guard treats that as a hard close-block → slice failure → phase failure.

## Suggested directions

1. On `restart_phase`, reconcile/clear per-task commit records for the restarted slices so they aren't carried across the re-fork citing pre-restart SHAs.
2. The close guard could treat an unreachable commit whose content is already present on the integration branch (re-derived by the fresh run) as satisfied, rather than a hard block — or distinguish `role=unassigned` orphans from genuine lost work.
3. Reconsider whether a single slice's close-block should be recoverable in place (re-run just that slice's close) instead of failing the slice and cascading the whole phase.

## Repro context

- Pipeline `issue-3312`, implement phase, build `42eb9511f`, 19-slice sequential chain.
- Occurred after a `restart_phase` (prior run parked on quota streak-10).
- slice-2 integration branch `egg/issue-3312/slice-2`; orphan `task-2-4` commit `e1cf4a48208f89d54a404...` role=unassigned.
- Related: #3125 (closed; this is its guard firing again), #3337 (concurrent warm-resumed producer race in the same pipeline).
- Surfaced during the #3249 warm-resume measurement run.

--- a/orchestrator/contract_completeness.py
+++ b/orchestrator/contract_completeness.py
@@ -255,7 +255,11 @@ def evidence_commits(
 
 
 def format_evidence_rows(rows: list[dict[str, Any]]) -> str:
-    """One-line-per-row summary for evidence-reachability messages."""
+    """One-line-per-row summary for evidence-reachability messages.
+
+    Every row from ``evidence_commits`` is role-bound (``task.role`` is
+    truthy), so ``role`` is always a concrete producer role here.
+    """
     return "; ".join(
-        f"{r['id']} (role={r['role'] or 'unassigned'}, commit={r['commit']})" for r in rows
+        f"{r['id']} (role={r['role']}, commit={r['commit']})" for r in rows
     )

--- a/orchestrator/contract_completeness.py
+++ b/orchestrator/contract_completeness.py
@@ -215,9 +215,28 @@ def evidence_commits(
     before flipping status), and a cited-but-unreachable commit is a
     gap worth failing on either way.
 
-    Returns one dict per row (``id`` / ``role`` / ``commit``), empty
-    list when no row cites a commit, or ``None`` when ``slice_id`` was
-    given but no such slice exists (caller skips the gate).
+    **Role-less rows are excluded (#3339).** The gate exists to protect
+    the producer-scoped #3124 flow: a *confirmed producer* records a
+    post-confirmation commit that lives only on its local worktree
+    branch, so the deliverable would be lost on close. That producer
+    always owns a role. A task with ``role=None`` (planner left it
+    unassigned) has no producer obligated to push its commit through
+    ``consensus_push``, so a SHA it cites is bookkeeping — not a gated
+    deliverable. Yet a stray ``add-commit`` / ``complete-task --commit``
+    on such a row (and ``_merge_preserved_slice_runtime`` re-attaching
+    it across a ``restart_phase`` re-fork) made the gate fire on an
+    orphan commit ``role=unassigned`` for a slice that had *already
+    reached full BRC consensus* — every reviewer approved the diff that
+    is on the integration branch — and that single bookkeeping mismatch
+    failed the slice and cascaded the whole phase. Scoping the gate to
+    role-bound rows keeps its full protection for real producer commits
+    while no longer nuking a consensus-reached slice on an unassigned
+    orphan.
+
+    Returns one dict per role-bound row (``id`` / ``role`` /
+    ``commit``), empty list when no such row cites a commit, or ``None``
+    when ``slice_id`` was given but no such slice exists (caller skips
+    the gate).
 
     Row ordering is intentional: slice-declaration order outermost,
     task-declaration order within each slice. Callers (the close-merge
@@ -231,7 +250,7 @@ def evidence_commits(
         {"id": task.id, "role": task.role, "commit": task.commit}
         for sl in slices
         for task in sl.tasks or []
-        if task.commit
+        if task.commit and task.role
     ]
 
 

--- a/orchestrator/contract_completeness.py
+++ b/orchestrator/contract_completeness.py
@@ -260,6 +260,4 @@ def format_evidence_rows(rows: list[dict[str, Any]]) -> str:
     Every row from ``evidence_commits`` is role-bound (``task.role`` is
     truthy), so ``role`` is always a concrete producer role here.
     """
-    return "; ".join(
-        f"{r['id']} (role={r['role']}, commit={r['commit']})" for r in rows
-    )
+    return "; ".join(f"{r['id']} (role={r['role']}, commit={r['commit']})" for r in rows)

--- a/orchestrator/routes/pipelines.py
+++ b/orchestrator/routes/pipelines.py
@@ -11596,6 +11596,11 @@ def _check_slice_evidence_reachability(
     through the existing cascade + HITL escalation machinery instead of
     closing silently.
 
+    Only role-bound task rows are gated (#3339): the check exists for
+    the producer-scoped #3124 flow, so a ``role=unassigned`` row's
+    orphan commit is bookkeeping, not a gated deliverable, and must not
+    fail a consensus-reached slice. See ``cc.evidence_commits``.
+
     Failure posture mirrors the other completeness checks (#3081 /
     #3114): the gate degrades to ``None`` (close proceeds, warning
     logged) when the contract cannot be read, the slice id does not

--- a/orchestrator/tests/test_evidence_reachability_gate.py
+++ b/orchestrator/tests/test_evidence_reachability_gate.py
@@ -172,6 +172,29 @@ class TestEvidenceCommits:
         assert rows is not None
         assert rows[1] == {"id": "task-2-2", "role": "documenter", "commit": LATE_SHA}
 
+    def test_roleless_row_with_orphan_commit_excluded(self, tmp_path: Path) -> None:
+        # #3339: a task the planner left unassigned (role=None) can still
+        # acquire a commit (a stray add-commit, or _merge_preserved_slice_
+        # runtime re-attaching one across a restart_phase re-fork). That SHA
+        # is bookkeeping, not a producer deliverable, so the gate must not
+        # see it — otherwise an orphan role=unassigned commit fails a
+        # consensus-reached slice and cascades the whole phase.
+        contract = _contract_dict()
+        slice2 = contract["phases"][1]
+        roleless = slice2["tasks"][3]
+        assert roleless["id"] == "task-2-4"
+        assert "role" not in roleless  # planner left it unassigned
+        roleless["commit"] = "e" * 40  # the orphan SHA from #3339
+        contracts_dir = tmp_path / ".egg-state" / "contracts"
+        contracts_dir.mkdir(parents=True, exist_ok=True)
+        (contracts_dir / f"{PIPELINE_ID}.json").write_text(json.dumps(contract))
+
+        rows = cc.evidence_commits(_load(tmp_path), "slice-2")
+        assert rows is not None
+        ids = [r["id"] for r in rows]
+        assert "task-2-4" not in ids
+        assert ids == ["task-2-1", "task-2-2", "task-2-3"]
+
     def test_format_evidence_rows(self) -> None:
         text = cc.format_evidence_rows(
             [
@@ -482,6 +505,28 @@ class TestCheckSliceEvidenceReachability:
         assert LATE_SHA in result
         assert INTEGRATION_BRANCH in result
         assert cc.EVIDENCE_GATE_ENV_VAR in result
+
+    def test_roleless_orphan_commit_does_not_fail_slice(self, tmp_path: Path, gate_env) -> None:
+        # #3339 regression: a slice whose only commit-bearing task is a
+        # role=unassigned orphan must close cleanly. The gateway reports
+        # the orphan SHA unreachable, but the roleless row is never put to
+        # the probe, so the gate returns None instead of failing the
+        # consensus-reached slice and cascading the phase.
+        contract = _contract_dict()
+        slice2 = contract["phases"][1]
+        for task in slice2["tasks"]:
+            task.pop("commit", None)  # drop the producer rows' commits
+        slice2["tasks"][3]["commit"] = "e" * 40  # task-2-4, roleless orphan
+        contracts_dir = tmp_path / ".egg-state" / "contracts"
+        contracts_dir.mkdir(parents=True, exist_ok=True)
+        (contracts_dir / f"{PIPELINE_ID}.json").write_text(json.dumps(contract))
+
+        spawner = _spawner(["e" * 40])  # gateway would flag it unreachable
+        result = _check_slice_evidence_reachability(
+            PIPELINE_ID, spawner, tmp_path, "slice-2", INTEGRATION_BRANCH
+        )
+        assert result is None
+        spawner.gateway.find_unreachable_evidence_commits.assert_not_called()
 
     def test_pre_loaded_contract_skips_internal_load(self, tmp_path: Path, gate_env) -> None:
         """When the caller pre-loads the contract (the close-path

--- a/orchestrator/tests/test_evidence_reachability_gate.py
+++ b/orchestrator/tests/test_evidence_reachability_gate.py
@@ -196,14 +196,16 @@ class TestEvidenceCommits:
         assert ids == ["task-2-1", "task-2-2", "task-2-3"]
 
     def test_format_evidence_rows(self) -> None:
+        # #3339: evidence_commits only ever emits role-bound rows, so the
+        # formatter always sees a concrete producer role.
         text = cc.format_evidence_rows(
             [
                 {"id": "task-2-2", "role": "documenter", "commit": "abc1234"},
-                {"id": "task-2-4", "role": None, "commit": "def5678"},
+                {"id": "task-2-4", "role": "coder", "commit": "def5678"},
             ]
         )
         assert "task-2-2 (role=documenter, commit=abc1234)" in text
-        assert "task-2-4 (role=unassigned, commit=def5678)" in text
+        assert "task-2-4 (role=coder, commit=def5678)" in text
 
 
 class TestEvidenceGateEnabled:

